src/function/scalar/system/aggregate_export.cpp

```diff
@@ -0,0 +1,286 @@
+		auto state_idx = state_data.sel->get_index(i);
+		auto state_entry = &((string_t *)state_data.data)[state_idx];
+		auto target_ptr = (const char *)local_state.state_buffer.get() + aligned_state_size * i;
+
+		if (state_data.validity.RowIsValid(state_idx)) {
+			D_ASSERT(state_entry->GetSize() == bind_data.state_size);
+			memcpy((void *)target_ptr, state_entry->GetDataUnsafe(), bind_data.state_size);
+		} else {
+			// create a dummy state because finalize does not understand NULLs in its input
+			// we put the NULL back in explicitly below
+			bind_data.aggr.initialize((data_ptr_t)target_ptr);
+		}
+		state_vec_ptr[i] = (data_ptr_t)target_ptr;
+	}
+
+	bind_data.aggr.finalize(local_state.addresses, nullptr, result, input.size(), 0);
+
+	for (idx_t i = 0; i < input.size(); i++) {
+		auto state_idx = state_data.sel->get_index(i);
+		if (!state_data.validity.RowIsValid(state_idx)) {
+			FlatVector::SetNull(result, i, true);
+		}
+	}
+}
+
+static void AggregateStateCombine(DataChunk &input, ExpressionState &state_p, Vector &result) {
+	auto &bind_data = ExportAggregateBindData::GetFrom(state_p);
+	auto &local_state = (CombineState &)*((ExecuteFunctionState &)state_p).local_state;
+
+	D_ASSERT(bind_data.state_size == bind_data.aggr.state_size());
+
+	D_ASSERT(input.data.size() == 2);
+	D_ASSERT(input.data[0].GetType().id() == LogicalTypeId::AGGREGATE_STATE);
+	D_ASSERT(input.data[0].GetType() == input.data[1].GetType());
+	D_ASSERT(input.data[0].GetType() == result.GetType());
+
+	VectorData state0_data, state1_data;
+	input.data[0].Orrify(input.size(), state0_data);
+	input.data[1].Orrify(input.size(), state1_data);
+
+	auto result_ptr = FlatVector::GetData<string_t>(result);
+
+	for (idx_t i = 0; i < input.size(); i++) {
+		auto state0_idx = state0_data.sel->get_index(i);
+		auto state1_idx = state1_data.sel->get_index(i);
+
+		auto &state0 = ((string_t *)state0_data.data)[state0_idx];
+		auto &state1 = ((string_t *)state1_data.data)[state1_idx];
+
+		// if both are NULL, we return NULL. If either of them is not, the result is that one
+		if (!state0_data.validity.RowIsValid(state0_idx) && !state1_data.validity.RowIsValid(state1_idx)) {
+			FlatVector::SetNull(result, i, true);
+			continue;
+		}
+		if (state0_data.validity.RowIsValid(state0_idx) && !state1_data.validity.RowIsValid(state1_idx)) {
+			result_ptr[i] =
+			    StringVector::AddStringOrBlob(result, (const char *)state0.GetDataUnsafe(), bind_data.state_size);
+			continue;
+		}
+		if (!state0_data.validity.RowIsValid(state0_idx) && state1_data.validity.RowIsValid(state1_idx)) {
+			result_ptr[i] =
+			    StringVector::AddStringOrBlob(result, (const char *)state1.GetDataUnsafe(), bind_data.state_size);
+			continue;
+		}
+
+		// we actually have to combine
+		D_ASSERT(state0.GetSize() == bind_data.state_size);
+		D_ASSERT(state1.GetSize() == bind_data.state_size);
+
+		memcpy(local_state.state_buffer0.get(), state0.GetDataUnsafe(), bind_data.state_size);
```

Comment:
I think there is an assumption floating around here that states do not have heap pointer components? I'm not sure that is true in general.

Reply:
We disallow custom finalizers so that should take care of that no?